### ISSUE TITLE
vimrc support

### ISSUE THREAD

**Issue by bluc41** [zed-industries/zed#4456]
### Check for existing issues

- [X] Completed

### Describe the feature

Being able to point to a .vimrc file and have vim mode use those bindings

**Comment by avi-cenna** [comment]
I've been using Zed and it's an awesome editor, but it seems that the keybinding features are quite limited right now.

Would it be easy to allow for chaining multiple commands? This way, people could replicate special functions with arrays of commands.

So for example, to comment out a paragraph, we would map `g c i p` to a sequence of commands like `[ "SelectParagraph", "Comment", "Unselect" ]`. 

If we could map a keybinding to an arbitrary sequence of keypresses that might work as well.

**Comment by ConradIrwin** [comment]
Yup! This is tracked at #7033. If you’d like to work on it with me, feel free to book time at https://calendly.com/conradirwin/pairing

**Comment by giorgioskij** [comment]
Hi! I too would love to have a configurable `.vimrc`: the main thing I miss is `jk` and `kj` to go back to normal mode, `U` mapped to `redo` and stuff like that. I'm writing on this thread because #7033 is not really the same thing, I'm note sure whether the two things can be considered a single feature or two separate ones. If you guys agree that this is useful, I would love to work on this.

**Comment by giorgioskij** [comment]
Nevermind, I just found out that I can actually do what I want with zed keybindings (https://zed.dev/docs/vim). Sorry for not reading docs earlier, with the issue being open I thought that this wouldn't be possible.

**Comment by awvalenti** [comment]
I'd love to be able to use my vimrc file, too. Some of my most important stuff:
```vim
nnoremap U <C-r>
noremap ' `
noremap - ^
nnoremap = $
vnoremap = $h
onoremap = $
noremap H <C-b>
noremap M zz
noremap L <C-f>
noremap _ g~l
nnoremap <Tab> "_
imap jk <Esc>
nnoremap K kJ
set clipboard=unnamed,unnamedplus
```

**Comment by betoharres** [comment]
it's really hard to move from vim/neovim if it's not something like gvim inside the editor; until that happens I'm going to stick with just good old vim

**Comment by t-var-s** [comment]
Maybe implement this through an LLM? 😅 It would help onboarding, specially if the process was transparent so the user could see what vimrc lines can do what in Zed.

**Comment by vhp1360** [comment]
Hello @ConradIrwin 
first of all,thanks for this very powerfull open source and free editor, near 15 years I have looked for a vim base IDE that support window like other IDEs and totally support vim.
about vimrc supporting:
in vimrc we can define new function and use it, but I cant find it in Zed settings, for example I defined highlight_occurances_of_current_word function or a comment function(comment current line and go to next) and ... but we cant in Zed.
there are maybe many settings and key binding those cant cover in Zed(and it is common subject)  but if we can use vimrc or at least define function like vim, give more power to developers.

many thanks

**Comment by georgebax** [comment]
Thanks to everyone for all the awesome work you've put into this editor, it's been years since I felt such a snappy application with a well-defined design philosophy. Personally I've tried switching a couple of times, and not being able to import my vimrc has been a dealbreak for me - the main advantage of using vim is, after all, the portability it provides. 
I'm really hoping this issue will gain some traction, but for anyone that's also been struggling I just found: github.com/jellydn/zed-101-setup.

**Comment by georgebax** [comment]
Hi @ConradIrwin thanks for your reply, here's a part of my .vimrc: 
```
inoremap jk <Esc>
inoremap kj <Esc>
inoremap jj <Esc>
inoremap kk <Esc>

" surround word with curly braces (mostly to be used as template tbh)
map <F5> i{<Esc>ea}<Esc>

set whichwrap=b,s,<,>,h,l,[,]
set autoindent
set ai
set tabstop=4
set ls=2
set ruler
set scrolloff=5
set incsearch
set hlsearch
set belloff=all
set textwidth=80
set ignorecase smartcase
set clipboard=unnamed,unnamedplus
set number
" Macros
nnoremap <C-j> viw"+p
```

I didn't include some parts that are related to the easymotion plugin (which the JetBrains suite supports, but I could live without ;) ) and vim-surround. I can also try passing them through an LLM too later today.

**Comment by tankorsmash** [comment]
Thanks for the existing vim mode!

I'd love to have something like `nmap j gj` etc to allow for moving down visual/wrapped lines. This comes up a lot in the Agent panel, where it's narrow, so almost every line of input is wrapped anyway.

**Comment by t-var-s** [comment]
> Maybe implement this through an LLM? 😅 It would help onboarding, specially if the process was transparent so the user could see what vimrc lines can do what in Zed.

Just to let you known that I got back to trying Zed by just giving my vimrc to some LLM and asking it for the closest equivalent in Zed settings.
